viam-cartographer/cartographer_patches/carto.patch

```diff
@@ -0,0 +1,304 @@
+-    file(RELATIVE_PATH REL_FIL ${PROJECT_SOURCE_DIR} ${ABS_FIL})
+-    get_filename_component(DIR ${REL_FIL} DIRECTORY)
+-    get_filename_component(FIL_WE ${REL_FIL} NAME_WE)
+-
+-    list(APPEND ALL_GRPC_SERVICE_SRCS "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.cc")
+-    list(APPEND ALL_GRPC_SERVICE_HDRS "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.h")
+-
+-    add_custom_command(
+-      OUTPUT "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.cc"
+-             "${PROJECT_BINARY_DIR}/${DIR}/${FIL_WE}.pb.h"
+-      COMMAND  ${PROTOBUF_PROTOC_EXECUTABLE}
+-      ARGS --cpp_out  ${PROJECT_BINARY_DIR}
+-        -I ${PROJECT_SOURCE_DIR}
+-        ${ABS_FIL}
+-      DEPENDS ${ABS_FIL}
+-      COMMENT "Running C++ protocol buffer compiler on ${ABS_FIL}"
+-      VERBATIM
+-    )
+-  endforeach()
+-  set_source_files_properties(${ALL_GRPC_SERVICE_SRCS} ${ALL_GRPC_SERVICE_HDRS} PROPERTIES GENERATED TRUE)
+-  list(APPEND ALL_LIBRARY_HDRS ${ALL_GRPC_SERVICE_HDRS})
+-  list(APPEND ALL_LIBRARY_SRCS ${ALL_GRPC_SERVICE_SRCS})
+-endif()
+ set(INSTALL_GENERATED_HDRS ${ALL_PROTO_HDRS} ${ALL_GRPC_SERVICE_HDRS})
+ 
+ add_library(${PROJECT_NAME} STATIC ${ALL_LIBRARY_HDRS} ${ALL_LIBRARY_SRCS})
+@@ -190,46 +126,10 @@ configure_file(
+   ${PROJECT_SOURCE_DIR}/cartographer/common/config.h.cmake
+   ${PROJECT_BINARY_DIR}/cartographer/common/config.h)
+ 
+-google_binary(cartographer_autogenerate_ground_truth
+-  SRCS
+-    cartographer/ground_truth/autogenerate_ground_truth_main.cc
+-)
+-
+-google_binary(cartographer_compute_relations_metrics
+-  SRCS
+-    cartographer/ground_truth/compute_relations_metrics_main.cc
+-)
+-
+-google_binary(cartographer_pbstream
+-  SRCS
+-  cartographer/io/pbstream_main.cc
+-)
+-
+-google_binary(cartographer_print_configuration
+-  SRCS
+-  cartographer/common/print_configuration_main.cc
+-)
+-
+-if(${BUILD_GRPC})
+-  google_binary(cartographer_grpc_server
+-    SRCS
+-      cartographer/cloud/map_builder_server_main.cc
+-  )
+-  target_link_libraries(cartographer_grpc_server PUBLIC grpc++)
+-  target_link_libraries(cartographer_grpc_server PUBLIC async_grpc)
+-  if(${BUILD_PROMETHEUS})
+-    target_link_libraries(cartographer_grpc_server PUBLIC ${ZLIB_LIBRARIES})
+-    target_link_libraries(cartographer_grpc_server PUBLIC prometheus-cpp-core)
+-    target_link_libraries(cartographer_grpc_server PUBLIC prometheus-cpp-pull)
+-  endif()
+-endif()
+-
+ target_include_directories(${PROJECT_NAME} SYSTEM PUBLIC
+   "${EIGEN3_INCLUDE_DIR}")
+ target_link_libraries(${PROJECT_NAME} PUBLIC ${EIGEN3_LIBRARIES})
+ 
+-target_include_directories(${PROJECT_NAME} SYSTEM PUBLIC
+-  "${CERES_INCLUDE_DIRS}")
+ target_link_libraries(${PROJECT_NAME} PUBLIC ${CERES_LIBRARIES})
+ 
+ target_include_directories(${PROJECT_NAME} SYSTEM PUBLIC
+@@ -280,60 +180,19 @@ target_link_libraries(${PROJECT_NAME} PUBLIC ${PROTOBUF_LIBRARY}
+   absl::strings
+   absl::synchronization
+   absl::time
+-  absl::utility 
++  absl::utility
+ )
++if (TARGET absl::log_internal_check_op)
```

Comment:
[note to self] This doesn't exist in protobuf < 23.0 so we need to case on it

Reply:
More specifically, it didn't exist in some previous version of abseil. But that also means protobuf didn't USE it either. And thus this is a crude check... if it exists, assume protobuf needs it, as both packages probably got updated at the same time.